vessel-runtime/src/main/java/com/textnow/android/vessel/Profiler.kt

```diff
@@ -0,0 +1,358 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2020 TextNow, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.textnow.android.vessel
+
+import kotlinx.coroutines.CoroutineDispatcher
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.coroutines.coroutineContext
+import kotlin.system.measureTimeMillis
+
+/**
+ * A time span over which some database operation occurred.
+ * Used in profiling database operations when using a [Vessel] instance.
+ * This shows where time is being spent when using [Vessel]
+ */
+enum class Span {
+    READ_FROM_DB,
+    WRITE_TO_DB,
+    PRELOAD_FROM_DB,
+    DELETE_FROM_DB,
+    REPLACE_IN_DB,
+    CLEAR_DB;
+
+    /**
+     * Friendly name for each enum - ex, "read", "write", etc.
+     */
+    fun niceName(): String {
+        return this.name.substring(0, this.name.indexOf("_")).lowercase()
+    }
+}
+
+/**
+ * A significant event that occurred when using a [Vessel] instance.
+ * Used to profile important events such as cache hits.
+ */
+enum class Event {
```

Comment:
If this was a CacheHitEvent, with READ/WRITE/etc, you wouldn't need niceName()?

Reply:
It just keeps the summary output a bit nicer

Instead of:

```
Spans
           span | count |  time
   READ_FROM_DB |     2 |    56
    WRITE_TO_DB |    16 |   474
        PRELOAD |     1 |    17

Events
          event | count
 CACHE_HIT_READ |   313
CACHE_HIT_WRITE |   128
```

You have:

```
Database Operations
           span | count |  time
   read |     2 |    56
    write |    16 |   474
        preload |     1 |    17

Cache Hits
          event | count
 read |   313
write |   128
```

Reply:
A full report looks something like:

```
Database I/O times, sorted by time spent
    span │ count │ time 
═════════╪═══════╪══════
   write │    15 │  947 
    read │    17 │  815 
 preload │     1 │   18 
  delete │     0 │    0 
 replace │     0 │    0 
   clear │     0 │    0 
─────────┼───────┼──────
         │    33 │ 1780 

Cache hits, sorted by hit count
   event │ count 
═════════╪═══════
    read │   202 
   write │   134 
  delete │     0 
 replace │     0 
─────────┼───────
         │   336 

Database I/O times, by thread/coroutine, sorted by time spent
      type │                        name │ write │ read │ preload │ delete │ replace │ clear │ total 
═══════════╪═════════════════════════════╪═══════╪══════╪═════════╪════════╪═════════╪═══════╪═══════
 COROUTINE │              Dispatchers.IO │   827 │  197 │       0 │      0 │       0 │     0 │  1024 
    THREAD │  DefaultDispatcher-worker-2 │     0 │  263 │      18 │      0 │       0 │     0 │   281 
    THREAD │  DefaultDispatcher-worker-7 │     0 │  181 │       0 │      0 │       0 │     0 │   181 
    THREAD │                   Thread-38 │     0 │  109 │       0 │      0 │       0 │     0 │   109 
    THREAD │  DefaultDispatcher-worker-1 │   105 │    0 │       0 │      0 │       0 │     0 │   105 
    THREAD │  DefaultDispatcher-worker-8 │     0 │   38 │       0 │      0 │       0 │     0 │    38 
 COROUTINE │         Dispatchers.Default │    15 │   15 │       0 │      0 │       0 │     0 │    30 
    THREAD │                        main │     0 │    7 │       0 │      0 │       0 │     0 │     7 
    THREAD │  DefaultDispatcher-worker-4 │     0 │    3 │       0 │      0 │       0 │     0 │     3 
    THREAD │  DefaultDispatcher-worker-3 │     0 │    2 │       0 │      0 │       0 │     0 │     2 
    THREAD │                   Thread-35 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │                   Thread-55 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │                   Thread-39 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │ DefaultDispatcher-worker-12 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │                   Thread-48 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │  DefaultDispatcher-worker-5 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │  DefaultDispatcher-worker-6 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
───────────┼─────────────────────────────┼───────┼──────┼─────────┼────────┼─────────┼───────┼───────
           │                             │   947 │  815 │      18 │      0 │       0 │     0 │  1780 
```